// umberto/src/tasks/build-sdk.js
/**
 * @license Copyright (c) 2017-2023, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md.
 */

const upath = require( 'upath' );
const getSdkSources = require( '../sdk-builder/get-sdk-sources' );
const TemplateCollection = require( '../template/template-collection' );
const SdkBuilder = require( '../sdk-builder/sdk-builder' );

module.exports = config => {
	const templateCollection = new TemplateCollection( upath.join( config.themePath, 'layout/_sdk' ), {
		locals: config.templateLocals
	} );
	const sdkBuilder = new SdkBuilder( {
		BASE_PATH: config.BASE_PATH,
		startPage: '/index.html',
		sdkCollection: getSdkSources( upath.join( config.projectRootPath, config.projectPath, config.srcPath ) ),
		template: templateCollection,
		config: config.projectConfig,
		destinationPath: upath.join( config.outputDir, config.BASE_PATH ),
		canonicalUrlBeginning: config.canonicalUrlBeginning,
		macrosVariables: config.macrosVariables,
		docSearch: config.docSearch
	} );

	sdkBuilder.buildSdk();

	return sdkBuilder;
};
